### PR TITLE
Add width converter

### PR DESCRIPTION
**Pull request by tdb-alcorn**
This change implements a WidthConverter which can handle widths that are not strict multiple of each other. To achieve this we divide up data into blocks and then implement a double pointer queue with special handling for irregular offsets.

### CODE REVIEW

File: rtl/src/main/scala/tensil/util/decoupled/WidthConverter.scala [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    io.in.ready := !full
+    io.out.valid := !empty
+
+    io.out.bits := Cat(
+      (for (i <- 0 until blocksPerOutput)
+        yield arr(deqPtr + i.U)).reverse
+    )
+
+    // TODO handle pipe and flow for max throughput
```
Comment (petrohi) [inline rtl/src/main/scala/tensil/util/decoupled/WidthConverter.scala:87]: Two questions here.

What is the improvement we can see with pipe and flow handled?

Why code below is commented out?

**Review by petrohi — COMMENTED** [review]
Will approve once tested on FPGA.

File: rtl/src/main/scala/tensil/util/decoupled/WidthConverter.scala (hunk 1 above)
Comment (tdb-alcorn) [reply rtl/src/main/scala/tensil/util/decoupled/WidthConverter.scala:87]: I'll just delete it - these ideas come from the chisel queue implementation which I essentially copied and modified to implement this width converter. The advantage is that you could have the output be dequeued on the same cycle as the input, and the input could become ready on the same cycle the output becomes ready. The problem is that it's actually quite complicated to implement in the case of the width converter. It also doesn't really matter as long as we have queues on one or both sides of the converter.

**Comment by petrohi** [comment]
Everything checks out on FPGA. Also AXI Width Converter performance is matched almost exactly. LGTM!

  | ResNet20 CIFAR ONNX | YOLOv4 Tiny @416 ONNX | ResNet50 ImageNet ONNX
-- | -- | -- | --
A. ZCU104 baseline (ms) | 13.099 | 294.071 | 514.279
B. A and external AXI4S Width Converter instead of Transmisson | 11.168 | 262.804 | 473.956
C. B and two-cycle read/write fix | 9.661 | 214.892 | 415.953
D. C and fix for delay between ops | 7.387 | 120.748 | 256.987
E. D with WidthConverter instead of AXI4S Width Converter | 7.388 | 120.752 | 256.993

**Review by petrohi — APPROVED** [review]

**Pull request merged**